### PR TITLE
update `ty::TyKind` documentation

### PR DESCRIPTION
**Pull request by lcnr**
slightly unsure about `ty::Opaque` and `ty::Bound`/`ty::Placeholder`.

r? @jackh726 @nikomatsakis @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/ty/sty.rs [hunk 1]
```diff
@@ -175,23 +204,42 @@ pub enum TyKind<'tcx> {
...
+    /// See the `rustc-dev-guide` for more details about
+    /// [higher-ranked trait bounds][1] and [canonical queries][2].
+    ///
+    /// [1]: https://rustc-dev-guide.rust-lang.org/traits/hrtb.html
+    /// [2]: https://rustc-dev-guide.rust-lang.org/traits/canonical-queries.html
     Bound(ty::DebruijnIndex, BoundTy),
 
-    /// A placeholder type - universally quantified higher-ranked type.
+    /// A placeholder type, used during higher ranked subtyping to instantiate
+    /// bound variables.
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/sty.rs:234]: is that all placeholders are used for?

Comment (oli-obk) [reply compiler/rustc_middle/src/ty/sty.rs:234]: I think they are also used while traversing query boundaries (inference vars are converted to them, then the query is run, inside the query they are turned back into new inference vars. When returning a value from the query, inference vars are again turned into placeholders, once the query has returned, those placeholders are turned back into to the original inference vars)

**Comment by nikomatsakis** [comment]
Thanks for doing this, @lcnr, these kinds of things are always helpful.

File: compiler/rustc_middle/src/ty/sty.rs [hunk 2]
```diff
@@ -175,23 +204,42 @@ pub enum TyKind<'tcx> {
...
     /// Opaque (`impl Trait`) type found in a return type.
+    ///
     /// The `DefId` comes either from
     /// * the `impl Trait` ast::Ty node,
     /// * or the `type Foo = impl Trait` declaration
-    /// The substitutions are for the generics of the function in question.
-    /// After typeck, the concrete type can be found in the `types` map.
+    ///
+    /// For RPIT the substitutions are for the generics of the function,
+    /// while for TAIT it is used for the generic parameters of the alias.
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/sty.rs:213]:
```suggestion
    /// while for TAIT it is used for the generic parameters of the alias.
    ///
    /// During codegen, `type_of(def_id)` can be used to get the underlying type
```

File: compiler/rustc_middle/src/ty/sty.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_middle/src/ty/sty.rs:236]: > think they are also used while traversing query boundaries (inference vars are converted to them, then the query is run, inside the query they are turned back into new inference vars.

don't we convert them to `ty::Bound`?

> Currently they are used for that, they represent "some unknown type".

"that" being "used during higher ranked subtyping to instantiate bound variables"?

**Comment by jackh726** [comment]
This is fantastic. Thanks!

@bors r+

**Pull request merged**